ariadne/cardano/src/Ariadne/MainTemplate.hs

```diff
@@ -154,7 +156,23 @@ initializeEverything MainSettings {..}
     serviceAction :: IO ()
     serviceAction =
       raceWithUpdateCheckAction $
-      uiAction `race_`
       cardanoAction
 
-  return $ withAsync initAction $ \_ -> serviceAction
+    mainAction :: IO ()
+    mainAction = do
+      -- First launch init action and wait until it completes
+      -- TODO is async really neded for that?
+      wait =<< async initAction
+      -- Then spawn backend actions in async thread
+      serviceThread <- async serviceAction
```

Comment:
So the only reason we had the problem in the first place is that we run uiAction with race in serviceAction?

Do I understand right that you propose to run init and backend using withAsync and then run uiAction without any async in the current thread?

Reply:
> So the only reason we had the problem in the first place is that we run uiAction with race in serviceAction?

AFAIU, yes, because according to the docs it's equivalent to
```
race left right =
  withAsync left $ \a ->
  withAsync right $ \b ->
  waitEither a b
```

> Do I understand right that you propose to run init and backend using withAsync and then run uiAction without any async in the current thread?

I propose something like
```
  initAction
  withAsync serviceAction \serviceThread -> do
    link serviceThread
    uiAction
```